### PR TITLE
Fix BlazorWebView disposal

### LINKED ISSUE #7277: Dispose() on Razor component not called on application close (Blazor .MAUI Windows app)
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Dispose-on-Razor-component-not-called-/10043254)._

---
[severity:I'm unable to use this version]
I am having serious component-lifetime issue. The problem is with implementing Dispose from IDisposable on a razor component. The method is perfectly called one the page gets off from viewing, ex. when switching to other view. Also, the OnInitialized() is being called. But, once clicking the [x] close of the entire application, the Dispose() does not happen. As a result null-reference exceptions appear whenever the background tasks not being stop before application close try to access not existing objects. It is not much seen in example below, but it happens in a larger application when a timer is hit during application close.

How to reproduce:
1) Create a new application from template
2) Add a System.Threading.Timer object to SurveyPrompt.razor and init it in OnInitialized()
3) add @implements IDisposable
4) Add Dispose() => timer?.Dispose();
5) Run the application
6) The Dispose() / OnInitialized() works perfectly when switching between Home/Counter menu items, but when closing application with top-right [X] the Dispose() is not being hit.

IMHO this issue prevents MAUI to go public, I cannot control background tasks and cannot dispose resources properly.

---
#### Feedback Bot on 5/16/2022, 01:47 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by MackinnonBuck**
### Description of Change

~~`BlazorWebViewHandler.DisconnectHandler()` has not been getting invoked when the `BlazorWebView` gets unloaded. This was due to the incorrect assumption that `DisconnectHandler()` gets invoked automatically by the framework (see #3604). The result is that none of the `*WebViewManager` disposal logic was running, including the logic that ultimately runs `Dispose()` in Blazor components. This PR solves this problem in addition to a bug preventing disposal on Windows when the application closes.~~

Fixes an issue where `DisconnectHandler()` throws an exception on Android due to the double-closing of Android WebView message channel ports.

### Issues Fixed

~~Fixes 7277~~

Repurposed to fix a potential future bug where `DisconnectHandler()` throws an exception on Android.

### CODE REVIEW

File: src/BlazorWebView/src/Maui/Android/AndroidWebKitWebViewManager.cs [hunk 1]
```diff
@@ -88,19 +88,6 @@ internal void SetUpMessageChannel()
...
-
-			if (_nativeToJSPorts is not null)
-			{
-				foreach (var port in _nativeToJSPorts)
-				{
-					port?.Close();
-				}
-			}
-		}
-
```
Comment (MackinnonBuck) [inline src/BlazorWebView/src/Maui/Android/AndroidWebKitWebViewManager.cs:103]: Turns out that these ports get closed automatically when the WebView gets unloaded. Now that this logic was running for the first time, an exception was being thrown because the ports were being closed twice. Hence the removal of this method.

File: src/BlazorWebView/src/Maui/Android/AndroidWebKitWebViewManager.cs [hunk 2]
```diff
@@ -88,19 +88,6 @@ internal void SetUpMessageChannel()
 			_webview.PostWebMessage(new WebMessage("capturePort", destPort), AndroidAppOriginUri);
 		}
 
-		protected override async ValueTask DisposeAsyncCore()
-		{
-			await base.DisposeAsyncCore();
-
-			if (_nativeToJSPorts is not null)
```
Comment (TanayParikh) [inline src/BlazorWebView/src/Maui/Android/AndroidWebKitWebViewManager.cs:95]: I don't believe we need the global `_nativeToJSPorts` field anymore, can we make that a local var again?

**Review by TanayParikh — COMMENTED** [review]
Great find!

**Review by mkArtakMSFT — APPROVED** [review]
Thanks @MackinnonBuck.
Don't we need unit-tests for these now?

**Review by TanayParikh — APPROVED** [review]

File: src/BlazorWebView/src/Maui/BlazorWebView.cs [hunk 3]
```diff
@@ -61,6 +67,17 @@ public virtual IFileProvider CreateFileProvider(string contentRootDir)
...
 
+#if ANDROID
+		[System.Runtime.Versioning.SupportedOSPlatform("android23.0")]
+#elif IOS
+		[System.Runtime.Versioning.SupportedOSPlatform("ios11.0")]
+#endif
+		private void OnUnloaded(object? sender, EventArgs e)
+		{
+			// See: https://github.com/dotnet/maui/issues/3604
+			Handler?.DisconnectHandler();
```
Comment (javiercn) [inline src/BlazorWebView/src/Maui/BlazorWebView.cs:78]:
```suggestion
			Handler?.DisconnectHandler();
			Unloaded -= OnUnloaded;
```

**Review by javiercn — APPROVED** [review]

File: src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs [hunk 4]
```diff
@@ -24,7 +24,10 @@ protected override void OnPropertyChanged(string? propertyName = null)
 
 		private void Window_Destroying(object? sender, EventArgs e)
 		{
-			// see: https://github.com/microsoft/microsoft-ui-xaml/issues/6872
+			// See: https://github.com/dotnet/maui/issues/7277
+			Handler?.DisconnectHandler();
```
Comment (mattleibow) [inline src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @PureWeen will this cause issues at all?

Comment (Eilon) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @MackinnonBuck why is this specifically needed on Windows and not other platforms?

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: The original issue that this PR is solving was reported for Windows. On Android and iOS, the `Window.Destroying` event does not get invoked when the app gets closed (#6993), so it wouldn't make a difference anyway. That said, we could make this line of code platform-independent, but that might introduce a risk that it doesn't work as expected on Android and iOS if/when #6993 gets fixed. Do you think it's worth investigating another way to make this work on other platforms?

Comment (PureWeen) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: Each platform has a version of `Unloaded` if you want to tie into those? 
Android => ViewDetachedFromWindow
iOS => override WillMoveToWindow

It doesn't cover the window destroying case but it makes the behavior symmetrical for when a view is removed from the visual hierarchy.

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @PureWeen this PR already adds a handler for the `Unloaded` event to invoke `DisconnectHandler()`. Just so I understand, is your suggestion to use the platform-specific APIs instead?

Comment (Eilon) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: I guess the general question is: what is the recommended pattern for components/controls to dispose dependencies when they go away? Like say a media control has to dispose some native player component, or Blazor needs to dispose Razor objects, or an image control needs to dispose some file watcher component.

Comment (PureWeen) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: Aw sorry @MackinnonBuck I was thinking this was subscribing to `Unloaded` on the platform view. So in that case I think you'd want to subscribe to "loaded" inside `HandlerChanging` when a handler is first assigned. Unless a BlazorWebView is not reusable? Because there are cases when a user will keep around the reference to a `ContentPage` or possibly this `BlazorWebView`

> I guess the general question is: what is the recommended pattern for components/controls to dispose dependencies when they go away? Like say a media control has to dispose some native player component, or Blazor needs to dispose Razor objects, or an image control needs to dispose some file watcher component.

Currently we leave that up to the user based on what they want to do. For example, with a media control we can't really assume when a user wants to dispose of the platform components with a media control. 

Let's say a user wants to move a media control from one layout to another (switching to full screen, absolute, or fixed). In XF this was really hard to do because we would automatically dispose of the entire platform hierarchy. Same question I would say applies here as well. If a BlazorWebView is removed from the visual hierarchy does that mean you always want to dispose the platform components? For example, on Android if you have more than 3 tabs Android will remove your view from the visual hierarchy (so unloaded will fire) if it's not a tab that's to the left or right of the current tab. Is that ok here? Is this going to mess users up when using a BlazorWebView and a TabbedPage with more then 3 tabs? 

In a perfect world we we'd just rely on the GC so when the VirtualView/PlatformView are no longer referenced anywhere they'd just get cleaned up and finalized but I don't think this is going to work in 100 percent of the cases.

Adding some additional hooks for more information about when to disconnect/dispose will probably need to be added but at this point we leave the components life cycle up to the discretion of the component and hopefully the events that we do surface helps give enough information to make that decision. 

All that being said :-) Given the original issue maybe this can't really be fixed until we fix the `window` destroy life cycle events?

Comment (Eilon) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: Hmm, this is indeed more nuanced than I had considered. Is there any logical sense of "this things is definitely garbage and I definitely want to destroy everything", aside from waiting for GC, which is unpredictable? Or is this just not logically possible and it's truly up to each component to define a _heuristic_ (which I define to mean "provably incorrect logic that approximates truth") and work off of that?

Comment (PureWeen) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: Ignoring a possible failure of imagination, not really. 

A user might hold a reference to a Page for the entire life time of an app. For example, now with `Shell` since you can register pages as singleton a lot of people are going to register those pages as singletons which means every single control on that page will live for the entire life of the application. 

I think what we'll need to look at as we gather use cases are places where we can make good enough guesses and then let people influence those guesses.  For example, with Shell maybe we should always disconnect the handler if they've registered something as transient? I think one area here that would delight users is if they could scope to a page/navpage/window and then once that container is disposed that will naturally dispose of the children. We can't really do that currently with the limitations of MS.Ext.DI but just spit balling here :-) 

>  "this things is definitely garbage and I definitely want to destroy everything",
TLDR; not really but we should figure out a way to let users tap into life cycle points better.

Comment (PureWeen) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: https://github.com/dotnet/maui/issues/7381

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @PureWeen Since the primary use case of BlazorWebView will be MAUI Blazor apps where the WebView is basically the entire app, do you think a decent direction here would be to run cleanup logic when the PlatformView gets unloaded (which I think is what you were suggesting earlier)? It seems the alternative is to leave things the way they are where nothing ever gets disposed, which probably isn't ideal.

Comment (PureWeen) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @MackinnonBuck 

If the underlying handler is removed and then the same BlazorWebView instance gains a new handler later down the road. Does the BlazorWebView retain enough information about current state that the user would basically just start over from where they were? or would they go back to the beginning as if it was loaded for the first time?

Comment (pictos) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @PureWeen how about implementing a reference counter? If I'm not wrong `Comet` has this implemented. If that's a simple implementation we can add this to .NET MAUI and call the `DisconnectHandler` when we don't have anymore references at all.

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @PureWeen `BlazorWebView` retains hardly any state at all (it's managed by the underlying platform-specific WebView), so the user would basically be starting over from the beginning.

Comment (Eilon) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: > the primary use case of BlazorWebView will be MAUI Blazor apps where the WebView is basically the entire app

That might be the most common pattern, but certainly not the only one. I think _optimizing_ for the most common pattern can make sense, but we have to be careful to not cause significant harm to less-common scenarios (I'm not saying any suggestion here violates that, but I want to make sure we don't end up there). This stuff is very hard for me to think through 😄

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs:28]: @Eilon, if we compare MAUI Blazor with Blazor WebAssembly for a minute, we don't call `Dispose()` for Blazor WebAssembly components when, for example, the browser window is closed. If someone needs some logic to run when the application window is closing, couldn't you argue that the resource being disposed exists in a context broader than the BlazorWebView, so therefore the disposal logic should be handled from that broader context (in MAUI lifecycle events, for example)? For example, @PureWeen brought up offline that the issue faced in #7277 could be fixed by running the disposal logic from within a MAUI context. This obviously has the downside that if you want to share Razor components between supported Blazor platforms, you would potentially need multiple versions of the component that manage state/disposal logic differently. But this comes back to what I said earlier about there not really being an equivalent case in Blazor WebAssembly, so maybe that concern isn't valid?

Comment (MackinnonBuck) [reply src/BlazorWebView/src/Maui/Windows/BlazorWebView.Windows.cs]: The conclusion we reached offline is that we're going to hold off on having `BlazorWebView` determine when to call `DisconnectHandler()` for now. The existing behavior is more acceptable than initially thought, so we feel comfortable revisiting this topic later when the general concept of disposal in MAUI is better understood.

**Comment by MackinnonBuck** [comment]
We're holding off on manually invoking `DisconnectHandler()` for now. The primary change in this PR is now fixing the double-closing of message channel ports on Android for when `DisconnectHandler()` does get invoked.

**Review by Eilon — APPROVED** [review]
